docs/madr/decisions/015-retry-policy.md

```diff
@@ -0,0 +1,223 @@
+      - envoy_ratelimited
+      - retriable_4xx
+      - refused_stream
+      - retriable_status_codes
+      - retriable_headers
+      - http3_post_connect_failure
+  grpc:
+    numRetries: 5
+    perTryTimeout: 200ms
+    backOff:
+      baseInterval: 20ms
+      maxInterval: 1s
+    retryOn:
+      - cancelled
+      - deadline_exceeded
+      - internal
+      - resource_exhausted
+      - unavailable
+  tcp:
+    maxConnectAttempts: 3
+```
+
+We can slightly change the conf for better compatibility with merging:
+
+```yaml
+default:
+  http:
+    numRetries: 5
+    perTryTimeout: 200ms
+    backOff:
+      baseInterval: 20ms
+      maxInterval: 1s
+    retriableMethods:
+      - GET
+    retryOn:
+      - 5xx
+      - gateway_error
+      - reset
+      - connect_failure
+      - envoy_ratelimited
+      - retriable_4xx
+      - refused_stream
+      - http3_post_connect_failure
+    additionalRetriableStatusCodes:
+      - 500
+      - 504 
+    retriableHeaders:
+      - name: header-name
+        exact: value
+    retriableRequestHeaders:
+      - name: should-retry-this
+        exact: yes
+  grpc:
+    numRetries: 5
+    perTryTimeout: 200ms
+    backOff:
+      baseInterval: 20ms
+      maxInterval: 1s
+    retryOn:
+      - cancelled
+      - deadline_exceeded
+      - internal
+      - resource_exhausted
+      - unavailable
+  tcp:
+    maxConnectAttempts: 3
+```
+
+Changes:
+
+1. Rename "http.retryOn.all_5xx" to "http.retryOn.5xx". Apparently it was a limitation of proto enums.
+2. Rename "retriableStatusCodes" to "additionalRetriableStatusCodes" 
+because this field provides additional status codes to whatever already specified in "retryOn".
+Today there is a validation if you set "retriableStatusCodes" you must provide "retriable_status_codes" to the "retryOn" list. 
+Due to a multi-level organisation of policies now it's impossible to validate.
+I think the best option is to add "retriable_status_codes" to "retryOn" automatically if "additionalRetriableStatusCodes" is specified. 
+3. New field "retriableHeaders" (or "retriableResponseHeaders"). Even though we have already had "retryOn.retriable_headers" 
+there was no way to provide a list of headers. 
+If this field is set we automatically add "retriable_headers" to "retryOn".
+4. New field "retriableRequestHeaders". Purely for the sake of symmetry, 
+if we allow retrying on response headers why not allow that on request headers. 
+
+#### Other retry-related Envoy parameters
+
+So far we didn't get requests from users to extend the policy with other Envoy parameters.
+But there are few additional things we can potentially support.
+
+##### per_try_idle_timeout
+
+Same as "idle_timeout" but enforced on each individual attempt.
+
+##### Host selection
+
+Envoy has "host predicates" and "priority predicates" to enforce an additional logic 
+when selecting a host for the attempt:
+
+* don't retry on previously retried hosts
+* don't retry on canary hosts
```

Comment:
what is a canary host in this context?

Reply:
When you specify LbEndpoint for EDS you can mark a host with `canary: true`. We're not doing it at that moment

Reply:
OK, would be good to include that context because canaries only exist at the level of like a `TrafficRoute` so this is confusing IMO